src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PostMergeBlockProducer.cs

```diff
@@ -79,22 +80,29 @@ public virtual Block PrepareEmptyBlock(BlockHeader parent, PayloadAttributes? pa
         protected override Block PrepareBlock(BlockHeader parent, PayloadAttributes? payloadAttributes = null)
         {
             Block block = base.PrepareBlock(parent, payloadAttributes);
-            AmendHeader(block.Header);
+            AmendHeader(block.Header, parent);
             return block;
         }
 
         protected override BlockHeader PrepareBlockHeader(BlockHeader parent, PayloadAttributes? payloadAttributes = null)
         {
             BlockHeader blockHeader = base.PrepareBlockHeader(parent, payloadAttributes);
-            AmendHeader(blockHeader);
+            AmendHeader(blockHeader, parent);
             return blockHeader;
         }
 
         // TODO: this seems to me that it should be done in the Eth2 seal engine?
-        private void AmendHeader(BlockHeader blockHeader)
+        private void AmendHeader(BlockHeader blockHeader, BlockHeader parent)
         {
             blockHeader.ExtraData = _blocksConfig.GetExtraDataBytes();
             blockHeader.IsPostMerge = true;
+            IReleaseSpec spec = _specProvider.GetSpec(blockHeader);
+
+            if (spec.IsEip4844Enabled)
+            {
```

Comment:
This is for empty blocks only. Do we set DataGasUsed and ExcessDataGas for nonempty blocks correctly (blocks with blobs)? Do we have an engine or hive test for that?

Reply:
Yes, we set it later. Hive tests cover it well. I guess we can also add data gas integration tests to the existing unit tests, adding it

Reply:
Cool. I'm marking it as resolved